misc/mdns/src/service.rs

```diff
@@ -278,48 +300,43 @@ impl fmt::Debug for MdnsService {
 
 /// A valid mDNS packet received by the service.
 #[derive(Debug)]
-pub enum MdnsPacket<'a> {
+pub enum MdnsPacket {
     /// A query made by a remote.
-    Query(MdnsQuery<'a>),
+    Query(MdnsQuery),
     /// A response sent by a remote in response to one of our queries.
-    Response(MdnsResponse<'a>),
+    Response(MdnsResponse),
     /// A request for service discovery.
-    ServiceDiscovery(MdnsServiceDiscovery<'a>),
+    ServiceDiscovery(MdnsServiceDiscovery),
 }
 
 /// A received mDNS query.
-pub struct MdnsQuery<'a> {
+pub struct MdnsQuery {
     /// Sender of the address.
     from: SocketAddr,
     /// Id of the received DNS query. We need to pass this ID back in the results.
     query_id: u16,
-    /// Queue of pending buffers.
-    send_buffers: &'a mut Vec<Vec<u8>>,
 }
 
-impl<'a> MdnsQuery<'a> {
-    /// Respond to the query.
+impl MdnsQuery {
+    /// Build query response.
     ///
     /// Pass the ID of the local peer, and the list of addresses we're listening on.
     ///
     /// If there are more than 2^16-1 addresses, ignores the others.
     ///
     /// > **Note**: Keep in mind that we will also receive this response in an `MdnsResponse`.
     #[inline]
-    pub fn respond<TAddresses>(
+    pub fn build_response<TAddresses>(
         self,
         peer_id: PeerId,
         addresses: TAddresses,
         ttl: Duration,
-    ) -> Result<(), MdnsResponseError>
+    ) -> Result<Vec<u8>, MdnsResponseError>
```

Comment:
Well, that's not great.
The `respond` method was a way to tell **the service** to send a response.
Now that `MdnsQuery` is detached from the service, it doesn't make sense to keep the API similar to what it was before.

Reply:
Can you add some more details here? Due to  the fact that `MdnsQuery` does not have  a reference to the `send_buffer` anymore it only builds the response, which later on is send by the `Mdns` `NetworkBehaviour` implementation to the `MdnsService`.

Reply:
What I mean is that it doesn't make sense for the `MdnsQuery` object itself to build the response. With the exception of `query_id`, the `self` isn't necessary.